### DISCUSSION TITLE
Get Master ID in Details Form

### DISCUSSION THREAD

**Discussion by rhemasystems** [serenity-is/Serenity#6040] [Help]
I have a form which when its open has two tabs on the form Project (Header) and Stages (Details)
How can I pass the header id when I click add button to add new details 
Below is the header 
![image](https://user-images.githubusercontent.com/50826372/137540143-edae5053-8663-430a-90a0-7685884eb598.png)

Below is the details Tab, when I click on the New button I want the projectid record of the header to display on the details form, it shows for existing records but not new records as seen in the last image
![image](https://user-images.githubusercontent.com/50826372/137539803-3a511b3b-df5f-4a49-818c-d7336547d4fe.png)

Editing a record

![image](https://user-images.githubusercontent.com/50826372/137540959-319741cb-d923-4776-b2c5-cb8360d862de.png)

New record

![image](https://user-images.githubusercontent.com/50826372/137542096-4a527672-5646-44c5-bcbd-544a37cb0599.png)

**Comment by VictorTomaili** [comment]
you can use `loadEntityAndOpenDialog`

**Reply by rhemasystems** [reply]
I have the below in the details Dialog, but how to get the master id and pass it to the cascadeValue value is my issue

`this.form.ProjectId.cascadeField = ProjectsRow.Fields.Id;`
`protected beforeLoadEntity(entity) {`
      `          super.beforeLoadEntity(entity);`
`          this.form.ProjectId.cascadeValue =??? ;//`
`}`

**Reply by go2ismail** [reply]
Hello,
The easiest way would be by getting the information from URL, example:
![image](https://user-images.githubusercontent.com/2275051/137626132-aa230884-ecd3-4909-8566-1bb6dcdd3b1b.png)

then on your XYZDetailDialog.ts you can do kind of:

`        protected afterLoadEntity() {
            super.afterLoadEntity();

            if (this.isNew()) {

                var url = window.location.href;
                url = url.substr(url.indexOf('#') + 1);
                var parts = url.split('/');

                if (parts[0] == 'edit') {
                    var masterId = parts[1];
                    this.form.ProjectId.cascadeValue = masterId;
                }

            }

        }`

Note:
Small note for us, master-detail already handled nicely by serenity using MasterDetailRelation attribute, no need to add additional task for getting master-detail to work. This approach, getting masterid at DetailDialog.ts, maybe needed for special occasion such as displaying the the master lookup control at DetailDialog.

**Reply by rhemasystems** [reply]
Thank you so much, the filter worked.
Quick one since the filtered record will always result in a single record being left in the drop down list, how can i auto select the record and disable the drop down after loading the form?

**Reply by go2ismail** [reply]
Hello,

To auto select the record, at afterLoadEntity you can select the record, maybe like below:
`
protected afterLoadEntity(){
   super.afterLoadEntity();
   var selectedId = xx;//fill with your desired Id
   this.form.YourDDL.value = selectedId;

   //to disable the control you can access its element, like below:
   this.form.YourDDL.element.prop('disabled', true);
}
`

**Comment by luongnv6** [comment]
Example from demo
https://github.com/serenity-is/common-features/blob/master/src/Serenity.Demo.Northwind/Serenity.Demo.Northwind/Customer/CustomerOrdersGrid.ts
